core/Util/plJPEG.cpp

```diff
@@ -195,7 +266,84 @@ void plJPEG::DecompressJPEG(hsStream* S, void* buf, size_t size)
     jpeg_finish_decompress(&ji.dinfo);
 }
 
-void plJPEG::CompressJPEG(hsStream* S, void* buf, size_t size)
+
+plMipmap* plJPEG::DecompressJPEG(hsStream* S)
+{
+    plJPEG& ji = Instance();
+
+    jpeg_hsStream_src(&ji.dinfo, S);
+    jpeg_read_header(&ji.dinfo, TRUE);
+    ji.dinfo.out_color_space = JCS_EXT_BGRA;    // Data stored as RGB on disk but Plasma uses BGR
+    jpeg_start_decompress(&ji.dinfo);
+
+    int row_stride = ji.dinfo.output_width * ji.dinfo.out_color_components;
+    int out_stride = row_stride;
+    RAII_JSAMPROW<1> jbuffer(row_stride);
+
+    // Start with a reasonable size for the buffer
+    std::vector<uint8_t> buffer(ji.dinfo.output_width * ji.dinfo.output_height * ji.dinfo.out_color_components);
+
+    size_t offs = 0;
+    while (ji.dinfo.output_scanline < ji.dinfo.output_height) {
+        while (offs + out_stride > buffer.capacity()) {
+            buffer.reserve(buffer.capacity() + INPUT_BUF_SIZE);
+        }
+        jpeg_read_scanlines(&ji.dinfo, jbuffer.data, 1);
+        for (size_t x = 0; x < ji.dinfo.output_width; x++) {
+            memcpy(((unsigned char*)buffer.data()) + offs + (x * 4),
+                jbuffer.data[0] + (x * ji.dinfo.out_color_components),
+                ji.dinfo.out_color_components);
+        }
+        offs += out_stride;
+    }
+
+    jpeg_finish_decompress(&ji.dinfo);
+    buffer.shrink_to_fit();
+
+    plMipmap* newMipmap = new plMipmap(ji.dinfo.output_width, ji.dinfo.output_height, 1, plMipmap::kUncompressed, plMipmap::kRGB8888);
+    newMipmap->setImageData(buffer.data(), buffer.size());
```

Comment:
```suggestion
    newMipmap->setImageData(buffer.get(), ji.dinfo.output_width * ji.dinfo.output_height * ji.dinfo.out_color_components);
```

You may want to store the size of the buffer as a constant to avoid repeating that,